### PR TITLE
[osh] Implement BashAssoc `assoc+=()`

### PR DESCRIPTION
**Pull request by akinomyoga**
https://oilshell.zulipchat.com/#narrow/channel/121539-oil-dev/topic/2024-11-22.20Meeting/near/485167138

### CODE REVIEW

File: osh/cmd_eval.py [hunk 1]
```diff
@@ -231,8 +231,19 @@ def PlusEquals(old_val, val):
                 raise AssertionError()  # parsing should prevent this
 
         elif case(value_e.BashAssoc):
-            # TODO: Could try to match bash, it will append to ${A[0]}
-            pass
+            if tag == value_e.Str:
+                e_die("Can't append string to array")
```
Comment (akinomyoga) [inline osh/cmd_eval.py:235]:
```suggestion
                e_die("Can't append string to associative arrays")
```

Oh, the error message should be "associative arrays"...

File: osh/cmd_eval.py [hunk 2]
```diff
@@ -231,8 +231,19 @@ def PlusEquals(old_val, val):
...
 
         elif case(value_e.BashAssoc):
-            # TODO: Could try to match bash, it will append to ${A[0]}
-            pass
+            if tag == value_e.Str:
+                e_die("Can't append string to associative arrays")
+
+            elif tag == value_e.BashAssoc:
+                assoc_lhs = cast(value.BashAssoc, UP_old_val)
+                assoc_rhs = cast(value.BashAssoc, UP_val)
```
Comment (andychu) [inline osh/cmd_eval.py:239]: Do we need to consider `tag == value_e.BashArray` ?


```
assoc+=([k]=v)  # right now we consider this BashAssoc

assoc+=(x y) # I think we consider this an array

assoc+=str
```


I guess we can add an assert for now?

Comment (akinomyoga) [reply osh/cmd_eval.py:239]: > Do we need to consider `tag == value_e.BashArray` ?

I just followed the current code for BashArray.

https://github.com/oils-for-unix/oils/blob/0df89bf3d19c70a0003db71881c45e8637199ef5/osh/cmd_eval.py#L216-L230

File: osh/cmd_eval.py [hunk 3]
```diff
@@ -231,8 +231,19 @@ def PlusEquals(old_val, val):
...
+
+                for key in assoc_rhs.d.keys():
+                    assoc_lhs.d[key] = assoc_rhs.d[key]
+                val = assoc_lhs
+
+            else:
+                raise AssertionError()  # parsing should prevent this
```
Comment (akinomyoga) [inline osh/cmd_eval.py:246]: > We could add an assert for now -- just disallow the combination

Yeah, but I think that that is this line.

**Comment by andychu** [comment]
Ah OK, good point!

**Comment by akinomyoga** [comment]
Thank you!

**Pull request merged**